### ISSUE TITLE
Refactor task_placer

### ISSUE THREAD

**Issue by egithinji** [tijlleenders/ZinZen-scheduler#157]
`task_placer.rs` is becoming increasingly difficult to read. I'd like to move the responsibility of assigning the initial slots to the tasks to a separate module, `slot_assigner.rs`.
So the flow will be `task_generator` -> `slot_assigner` -> `task_placer`.

After that we can still improve `task_placer` further.

**Comment by tijlleenders** [comment]
Agree it's time to simplify (or compartimentalize) this.

We're talking about the initial creation of slots for each Task based upon the constraints right? 

Isn't slot_generator a better fitting name then?

It can be called in the initialization of the Task?

**Comment by tijlleenders** [comment]
Since we have multiple constraints on the slots that are possible - and that each have different logic - would it make more sense to approach the slot_generator by starting with the full calendar to be calculated Slot and then reducing this by intelligently looping through all the constraints?

Breaking down (or poking holes) in the initial monolith slot allows multiple constraints witb different logic to do their jobs one by one, independently from each other. Going at it the other way would imply combining all constraints at once - which is too complex.

**Comment by egithinji** [comment]
> Agree it's time to simplify (or compartimentalize) this.
> 
> We're talking about the initial creation of slots for each Task based upon the constraints right?
> 
> Isn't slot_generator a better fitting name then?
> 
> It can be called in the initialization of the Task?

I find `slot_assigner` more descriptive of what's happening during this stage. I.e. from a monolithic calendar, we're assigning specific slots to each task. `Generate` seems to imply that we're creating stuff, but maybe it's just me!

**Comment by egithinji** [comment]
> Since we have multiple constraints on the slots that are possible - and that each have different logic - would it make more sense to approach the slot_generator by starting with the full calendar to be calculated Slot and then reducing this by intelligently looping through all the constraints?
> 
> Breaking down (or poking holes) in the initial monolith slot allows multiple constraints witb different logic to do their jobs one by one, independently from each other. Going at it the other way would imply combining all constraints at once - which is too complex.

This is an interesting approach. Will try and implement it and see which is more readable.

**Comment by tijlleenders** [comment]
> > Agree it's time to simplify (or compartimentalize) this.
> > 
> > We're talking about the initial creation of slots for each Task based upon the constraints right?
> > 
> > Isn't slot_generator a better fitting name then?
> > 
> > It can be called in the initialization of the Task?
> 
> I find `slot_assigner` more descriptive of what's happening during this stage. I.e. from a monolithic calendar, we're assigning specific slots to each task. `Generate` seems to imply that we're creating stuff, but maybe it's just me!

Yeah so as discussed in the PR comments I agree slot_assigner is a better name for doing it that way. I'd suggest however not to do it that way - but to assign slots to tasks during Task creation - all in one step.

**Comment by egithinji** [comment]
Okay to resolve this issue, I'm first going to just decouple the slot assignment functionality into its own module. Then after that i'll create a separate issue for the other point.
